File: src/pipeline/zip.test.ts
/**
 * Unit tests for src/pipeline/zip.ts
 *
 * Strategy: build real in-memory zips with AdmZip and write them to a
 * tmp directory so extractAndValidateZip gets an actual file path.
 * No mocking of AdmZip internals — tests stay close to real behaviour.
 */

import AdmZip from 'adm-zip';
import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { extractAndValidateZip, ZipValidationError, ExtractedImages } from './zip';

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/** 1×1 transparent PNG — smallest valid PNG bytes */
const TINY_PNG = Buffer.from(
  '89504e470d0a1a0a0000000d49484452000000010000000108060000001f15c4' +
    '890000000a49444154789c6260000000020001e221bc330000000049454e44ae426082',
  'hex',
);

function makeZip(files: Record<string, Buffer | null>): string {
  const zip = new AdmZip();
  for (const [name, content] of Object.entries(files)) {
    zip.addFile(name, content ?? TINY_PNG);
  }
  const tmp = fs.mkdtempSync(path.join(os.tmpdir(), 'zip-test-'));
  const zipPath = path.join(tmp, 'test.zip');
  zip.writeZip(zipPath);
  return zipPath;
}

function isError(r: ExtractedImages | ZipValidationError): r is ZipValidationError {
  return 'kind' in r;
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe('extractAndValidateZip', () => {
  let destDir: string;

  beforeEach(() => {
    destDir = fs.mkdtempSync(path.join(os.tmpdir(), 'zip-dest-'));
  });

  afterEach(() => {
    fs.rmSync(destDir, { recursive: true, force: true });
  });

  // ── Happy paths ──────────────────────────────────────────────────────────

  describe('valid zips — bare number naming (old style)', () => {
    it('accepts 1.png', () => {
      const zipPath = makeZip({ '1.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      const { files } = result as ExtractedImages;
      expect(files).toHaveLength(1);
      expect(path.basename(files[0])).toBe('1.png');
    });

    it('accepts 1.jpg, 2.jpg, 3.jpg in any zip order', () => {
      const zipPath = makeZip({ '3.jpg': null, '1.jpg': null, '2.jpg': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      const { files } = result as ExtractedImages;
      expect(files.map(f => path.basename(f))).toEqual(['1.jpg', '2.jpg', '3.jpg']);
    });

    it('accepts zero-padded numbers (01.jpeg)', () => {
      const zipPath = makeZip({ '01.jpeg': null, '02.jpeg': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      const { files } = result as ExtractedImages;
      expect(files).toHaveLength(2);
    });

    it('accepts a full set of 10 images', () => {
      const files: Record<string, null> = {};
      for (let i = 1; i <= 10; i++) files[`${i}.png`] = null;
      const zipPath = makeZip(files);
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      expect((result as ExtractedImages).files).toHaveLength(10);
    });
  });

  describe('valid zips — parenthesised prefix naming (new style)', () => {
    it('accepts "dives (1).png"', () => {
      const zipPath = makeZip({ 'dives (1).png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      const { files } = result as ExtractedImages;
      expect(files).toHaveLength(1);
    });

    it('accepts "balloons pop (1).jpg", "balloons pop (2).jpg"', () => {
      const zipPath = makeZip({
        'balloons pop (2).jpg': null,
        'balloons pop (1).jpg': null,
      });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      const { files } = result as ExtractedImages;
      // sorted by number, not alpha
      expect(files.map(f => path.basename(f))).toEqual([
        'balloons pop (1).jpg',
        'balloons pop (2).jpg',
      ]);
    });

    it('accepts mixed prefix styles in the same zip', () => {
      const zipPath = makeZip({ '1.png': null, 'slide (2).png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      expect((result as ExtractedImages).files).toHaveLength(2);
    });

    it('accepts multi-word prefix with numbers in the prefix ("trip 2024 (1).jpg")', () => {
      const zipPath = makeZip({ 'trip 2024 (1).jpg': null, 'trip 2024 (2).jpg': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      expect((result as ExtractedImages).files).toHaveLength(2);
    });
  });

  // ── Error paths ──────────────────────────────────────────────────────────

  describe('empty', () => {
    it('returns empty when zip has no image files', () => {
      const zipPath = makeZip({ 'readme.txt': Buffer.from('hello') });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('empty');
    });

    it('returns empty when zip has only macOS junk', () => {
      const zipPath = makeZip({
        '__MACOSX/._1.png': TINY_PNG,
        '.DS_Store': Buffer.from('junk'),
      });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('empty');
    });
  });

  describe('too_many', () => {
    it('returns too_many for 11 images', () => {
      const files: Record<string, null> = {};
      for (let i = 1; i <= 11; i++) files[`${i}.png`] = null;
      const zipPath = makeZip(files);
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('too_many');
    });
  });

  describe('wrong_naming', () => {
    it('rejects a file with no number at all ("cover.png")', () => {
      const zipPath = makeZip({ 'cover.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('wrong_naming');
    });

    it('rejects a file with number but wrong format ("slide-1.png")', () => {
      const zipPath = makeZip({ 'slide-1.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('wrong_naming');
    });
  });

  describe('missing_numbers', () => {
    it('rejects a gap in sequence (1, 3 — missing 2)', () => {
      const zipPath = makeZip({ '1.png': null, '3.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('missing_numbers');
    });

    it('rejects sequence not starting at 1 (2, 3)', () => {
      const zipPath = makeZip({ '2.png': null, '3.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('missing_numbers');
    });

    it('rejects gap with parenthesised names ("dives (1).png", "dives (3).png")', () => {
      const zipPath = makeZip({ 'dives (1).png': null, 'dives (3).png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(true);
      expect((result as ZipValidationError).kind).toBe('missing_numbers');
    });
  });

  // ── Extraction ───────────────────────────────────────────────────────────

  describe('extraction', () => {
    it('creates destDir if it does not exist', () => {
      const nonExistent = path.join(destDir, 'subdir');
      const zipPath = makeZip({ '1.png': null });
      extractAndValidateZip(zipPath, nonExistent);
      expect(fs.existsSync(nonExistent)).toBe(true);
    });

    it('returns absolute paths that exist on disk', () => {
      const zipPath = makeZip({ '1.png': null, '2.png': null });
      const result = extractAndValidateZip(zipPath, destDir);
      expect(isError(result)).toBe(false);
      for (const f of (result as ExtractedImages).files) {
        expect(path.isAbsolute(f)).toBe(true);
        expect(fs.existsSync(f)).toBe(true);
      }
    });
  });
});

File: go/internal/pipeline/zip_test.go
package pipeline

import (
	"archive/zip"
	"bytes"
	"errors"
	"os"
	"path/filepath"
	"testing"
)

func makeTestZip(t *testing.T, files map[string][]byte) string {
	t.Helper()
	buf := new(bytes.Buffer)
	w := zip.NewWriter(buf)

	for name, content := range files {
		f, err := w.Create(name)
		if err != nil {
			t.Fatal(err)
		}
		if content == nil {
			content = []byte("fake image data")
		}
		_, err = f.Write(content)
		if err != nil {
			t.Fatal(err)
		}
	}

	err := w.Close()
	if err != nil {
		t.Fatal(err)
	}

	tmpDir := t.TempDir()
	zipPath := filepath.Join(tmpDir, "test.zip")
	err = os.WriteFile(zipPath, buf.Bytes(), 0o644)
	if err != nil {
		t.Fatal(err)
	}

	return zipPath
}

func TestExtractAndValidateZip(t *testing.T) {
	t.Run("accepts 1.png", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"1.png": nil,
		})
		destDir := t.TempDir()
		files, err := ExtractAndValidateZip(zipPath, destDir)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(files) != 1 {
			t.Fatalf("expected 1 file, got %d", len(files))
		}
		if filepath.Base(files[0]) != "1.png" {
			t.Errorf("expected 1.png, got %s", filepath.Base(files[0]))
		}
	})

	t.Run("accepts sorted consecutive files", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"3.jpg": nil,
			"1.jpg": nil,
			"2.jpg": nil,
		})
		destDir := t.TempDir()
		files, err := ExtractAndValidateZip(zipPath, destDir)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(files) != 3 {
			t.Fatalf("expected 3 files, got %d", len(files))
		}
		expected := []string{"1.jpg", "2.jpg", "3.jpg"}
		for i, name := range expected {
			if filepath.Base(files[i]) != name {
				t.Errorf("expected %s at index %d, got %s", name, i, filepath.Base(files[i]))
			}
		}
	})

	t.Run("accepts zero padded names", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"01.jpeg": nil,
			"02.jpeg": nil,
		})
		destDir := t.TempDir()
		files, err := ExtractAndValidateZip(zipPath, destDir)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(files) != 2 {
			t.Fatalf("expected 2 files, got %d", len(files))
		}
	})

	t.Run("accepts new style parenthesized naming", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"dives (1).png": nil,
			"dives (2).png": nil,
		})
		destDir := t.TempDir()
		files, err := ExtractAndValidateZip(zipPath, destDir)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(files) != 2 {
			t.Fatalf("expected 2 files, got %d", len(files))
		}
		if filepath.Base(files[0]) != "dives (1).png" {
			t.Errorf("expected dives (1).png, got %s", filepath.Base(files[0]))
		}
	})

	t.Run("rejects empty zip", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{})
		destDir := t.TempDir()
		_, err := ExtractAndValidateZip(zipPath, destDir)
		if err == nil {
			t.Fatal("expected error, got nil")
		}
		var zipErr *ZipError
		if !errors.As(err, &zipErr) || zipErr.Kind != "empty" {
			t.Errorf("expected empty ZipError, got %v", err)
		}
	})

	t.Run("rejects too many images", func(t *testing.T) {
		filesMap := map[string][]byte{
			"1.png": nil, "2.png": nil, "3.png": nil, "4.png": nil, "5.png": nil,
			"6.png": nil, "7.png": nil, "8.png": nil, "9.png": nil, "10.png": nil,
			"11.png": nil,
		}
		zipPath := makeTestZip(t, filesMap)
		destDir := t.TempDir()
		_, err := ExtractAndValidateZip(zipPath, destDir)
		if err == nil {
			t.Fatal("expected error, got nil")
		}
		var zipErr *ZipError
		if !errors.As(err, &zipErr) || zipErr.Kind != "too_many" {
			t.Errorf("expected too_many ZipError, got %v", err)
		}
	})

	t.Run("rejects missing numbers in sequence", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"1.png": nil,
			"3.png": nil,
		})
		destDir := t.TempDir()
		_, err := ExtractAndValidateZip(zipPath, destDir)
		if err == nil {
			t.Fatal("expected error, got nil")
		}
		var zipErr *ZipError
		if !errors.As(err, &zipErr) || zipErr.Kind != "missing_numbers" {
			t.Errorf("expected missing_numbers ZipError, got %v", err)
		}
	})

	t.Run("rejects wrong naming pattern", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"hello.png": nil,
		})
		destDir := t.TempDir()
		_, err := ExtractAndValidateZip(zipPath, destDir)
		if err == nil {
			t.Fatal("expected error, got nil")
		}
		var zipErr *ZipError
		if !errors.As(err, &zipErr) || zipErr.Kind != "wrong_naming" {
			t.Errorf("expected wrong_naming ZipError, got %v", err)
		}
	})

	t.Run("ignores macos metadata and directories", func(t *testing.T) {
		zipPath := makeTestZip(t, map[string][]byte{
			"1.png":               nil,
			"__MACOSX/._1.png":    nil,
			".DS_Store":           nil,
			"sub/2.png":           nil,
			"ignored_non_img.txt": []byte("text"),
		})
		destDir := t.TempDir()
		files, err := ExtractAndValidateZip(zipPath, destDir)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if len(files) != 1 {
			t.Errorf("expected only 1 valid file, got %d", len(files))
		}
	})
}
